rewriter/Minitest.cc

```diff
@@ -215,18 +227,58 @@ ast::ExpressionPtr runUnderEach(core::MutableContext ctx, core::NameRef eachName
     return stmt;
 }
 
+bool isDestructuringArg(core::GlobalState &gs, const ast::MethodDef::ARGS_store &args, const ast::ExpressionPtr &expr) {
+    auto *local = ast::cast_tree<ast::UnresolvedIdent>(expr);
+    if (local == nullptr || local->kind != ast::UnresolvedIdent::Kind::Local) {
+        return false;
+    }
+
+    auto name = local->name;
+    if (name.kind() != core::NameKind::UNIQUE || name.dataUnique(gs)->original != core::Names::destructureArg()) {
+        return false;
+    }
+
+    return absl::c_find_if(args, [name](auto &argExpr) {
+               auto *arg = ast::cast_tree<ast::UnresolvedIdent>(argExpr);
+               return arg && arg->name == name;
+           }) != args.end();
+}
+
+// Destructuring blocks are added to method/block entries as an InsSeq where the method body is the distinguished final
+// statement, and the statements of the InsSeq are all individual InsSeq expressions that make up the destructuring for
+// a specific argument.
+bool isDestructuringInsSeq(core::GlobalState &gs, const ast::MethodDef::ARGS_store &args, ast::InsSeq *body) {
+    return absl::c_all_of(body->stats, [&gs, &args](auto &stat) {
+        auto *insSeq = ast::cast_tree<ast::InsSeq>(stat);
+        if (insSeq == nullptr) {
+            return false;
+        }
+
+        auto *assign = ast::cast_tree<ast::Assign>(insSeq->stats.front());
+        return assign && isDestructuringArg(gs, args, assign->rhs);
+    });
+}
+
 // this just walks the body of a `test_each` and tries to transform every statement
 ast::ExpressionPtr prepareTestEachBody(core::MutableContext ctx, core::NameRef eachName, ast::ExpressionPtr body,
-                                       ast::MethodDef::ARGS_store &args, ast::ExpressionPtr &iteratee) {
-    auto *bodySeq = ast::cast_tree<ast::InsSeq>(body);
-    if (bodySeq) {
+                                       ast::MethodDef::ARGS_store &args, ast::InsSeq::STATS_store destructuringStmts,
+                                       ast::ExpressionPtr &iteratee) {
+    if (auto *bodySeq = ast::cast_tree<ast::InsSeq>(body)) {
+        if (isDestructuringInsSeq(ctx, args, bodySeq)) {
+            ENFORCE(destructuringStmts.empty(), "Nested destructuring statements");
```

Comment:
i haven't looked carefully but judging from the text of the enforce, is it actually possible for a user to encounter this enforce? should this be a `beginError` instead?

Reply:
The implication of this enforce firing would be that the destructuring desugaring managed to fire twice, but without anything that introduced new names in the middle. I couldn't figure out a case that would trigger this, but it felt like a reasonable defensive measure against changes to the desugar pass in the future.

Reply:
when i left this comment i was thinking it was talking about situations like

```ruby
test_each(...) do |(x, y)|
  test_each(...) do |(a, b)|
  end
end
```

but after reading the code more carefully, i don't think that's what you meant by nested here.

Reply:
That's correct. The example that you wrote would have a send with a block between the destructuring statements added to the beginning of the outer block.